File: queue/queue.go
package queue

import (
	"database/sql"
	"encoding/json"
	"errors"
	"time"
)

// LaQueue represents a queue backed by SQLite
type LaQueue struct {
	db        *sql.DB
	queueName string
}

// QueueItem represents an item in the queue
type QueueItem struct {
	ID            int64      `json:"id"`
	QueueName     string     `json:"queue_name"`
	Payload       []byte     `json:"payload"`
	CreatedAt     time.Time  `json:"created_at"`
	ScheduledAt   time.Time  `json:"scheduled_at"`
	Status        string     `json:"status"`
	Attempts      int        `json:"attempts"`
	LastAttemptAt *time.Time `json:"last_attempt_at,omitempty"`
}

// New creates a new LaQueue instance
func New(db *sql.DB, queueName string) *LaQueue {
	return &LaQueue{
		db:        db,
		queueName: queueName,
	}
}

// Enqueue adds a new item to the queue
func (q *LaQueue) Enqueue(payload any) (int64, error) {
	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return 0, err
	}

	result, err := q.db.Exec(
		`INSERT INTO queue_items (queue_name, payload) VALUES (?, ?)`,
		q.queueName, payloadBytes,
	)
	if err != nil {
		return 0, err
	}

	return result.LastInsertId()
}

// EnqueueWithDelay adds a new item to the queue with a specified delay
func (q *LaQueue) EnqueueWithDelay(payload any, delay time.Duration) (int64, error) {
	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return 0, err
	}

	scheduledAt := time.Now().Add(delay)

	result, err := q.db.Exec(
		`INSERT INTO queue_items (queue_name, payload, scheduled_at) VALUES (?, ?, ?)`,
		q.queueName, payloadBytes, scheduledAt,
	)
	if err != nil {
		return 0, err
	}

	return result.LastInsertId()
}

// Dequeue retrieves and claims the next available item from the queue
func (q *LaQueue) Dequeue() (*QueueItem, error) {
	tx, err := q.db.Begin()
	if err != nil {
		return nil, err
	}
	defer tx.Rollback()

	var item QueueItem
	now := time.Now()

	err = tx.QueryRow(`
		SELECT id, queue_name, payload, created_at, scheduled_at, status, attempts, last_attempt_at
		FROM queue_items
		WHERE queue_name = ? AND status = 'pending' AND scheduled_at <= ?
		ORDER BY scheduled_at ASC
		LIMIT 1
	`, q.queueName, now).Scan(
		&item.ID, &item.QueueName, &item.Payload, &item.CreatedAt,
		&item.ScheduledAt, &item.Status, &item.Attempts, &item.LastAttemptAt,
	)
	if err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, nil // No items in queue
		}
		return nil, err
	}

	// Mark the item as processing
	_, err = tx.Exec(`
		UPDATE queue_items
		SET status = 'processing', attempts = attempts + 1, last_attempt_at = ?
		WHERE id = ? AND queue_name = ?
	`, now, item.ID, q.queueName)
	if err != nil {
		return nil, err
	}

	if err = tx.Commit(); err != nil {
		return nil, err
	}

	item.Status = "processing"
	item.Attempts++
	item.LastAttemptAt = &now

	return &item, nil
}

// Complete marks a queue item as completed
func (q *LaQueue) Complete(id int64) error {
	_, err := q.db.Exec(`
		UPDATE queue_items
		SET status = 'completed'
		WHERE id = ? AND queue_name = ?
	`, id, q.queueName)
	return err
}

// Fail marks a queue item as failed
func (q *LaQueue) Fail(id int64) error {
	_, err := q.db.Exec(`
		UPDATE queue_items
		SET status = 'failed'
		WHERE id = ? AND queue_name = ?
	`, id, q.queueName)
	return err
}

// RetryWithDelay reschedules a failed item with a delay
func (q *LaQueue) RetryWithDelay(id int64, delay time.Duration) error {
	scheduledAt := time.Now().Add(delay)
	_, err := q.db.Exec(`
		UPDATE queue_items
		SET status = 'pending', scheduled_at = ?
		WHERE id = ? AND queue_name = ?
	`, scheduledAt, id, q.queueName)
	return err
}

// Size returns the number of pending items in the queue
func (q *LaQueue) Size() (int, error) {
	var count int
	now := time.Now()
	err := q.db.QueryRow(`
		SELECT COUNT(*) FROM queue_items
		WHERE queue_name = ? AND status = 'pending' AND scheduled_at <= ?
	`, q.queueName, now).Scan(&count)
	return count, err
}

File: ts/src/queue.ts
import Database from 'better-sqlite3';

export interface QueueItem {
  id: number;
  queueName: string;
  payload: Buffer;
  createdAt: Date;
  scheduledAt: Date;
  status: 'pending' | 'processing' | 'completed' | 'failed';
  attempts: number;
  lastAttemptAt: Date | null;
}

interface DBQueueItem {
  id: number;
  queue_name: string;
  payload: Buffer;
  created_at: string;
  scheduled_at: string;
  status: 'pending' | 'processing' | 'completed' | 'failed';
  attempts: number;
  last_attempt_at: string | null;
}

export class LaQueue {
  private db: Database.Database;
  private queueName: string;

  constructor(db: Database.Database, queueName: string) {
    this.db = db;
    this.queueName = queueName;
  }

  async enqueue(payload: any): Promise<number> {
    const payloadBuffer = Buffer.from(JSON.stringify(payload));
    const stmt = this.db.prepare(
      'INSERT INTO queue_items (queue_name, payload) VALUES (?, ?)'
    );
    const result = stmt.run(this.queueName, payloadBuffer);
    return result.lastInsertRowid as number;
  }

  async enqueueWithDelay(payload: any, delay: number): Promise<number> {
    const payloadBuffer = Buffer.from(JSON.stringify(payload));
    const scheduledAt = new Date(Date.now() + delay);
    const stmt = this.db.prepare(
      'INSERT INTO queue_items (queue_name, payload, scheduled_at) VALUES (?, ?, ?)'
    );
    const result = stmt.run(this.queueName, payloadBuffer, scheduledAt.toISOString());
    return result.lastInsertRowid as number;
  }

  dequeue(): QueueItem | null {
    const now = new Date().toISOString();

    return this.db.transaction(() => {
      const stmt = this.db.prepare(`
        SELECT id, queue_name, payload, created_at, scheduled_at, status, attempts, last_attempt_at
        FROM queue_items
        WHERE queue_name = ? AND status = 'pending' AND scheduled_at <= ?
        ORDER BY scheduled_at ASC
        LIMIT 1
      `);

      const item = stmt.get(this.queueName, now) as DBQueueItem;

      if (!item) {
        return null;
      }

      const updateStmt = this.db.prepare(`
        UPDATE queue_items
        SET status = 'processing', attempts = attempts + 1, last_attempt_at = ?
        WHERE id = ? AND queue_name = ?
      `);

      updateStmt.run(now, item.id, this.queueName);

      return {
        id: item.id,
        queueName: item.queue_name,
        payload: item.payload,
        createdAt: new Date(item.created_at),
        scheduledAt: new Date(item.scheduled_at),
        status: 'processing' as const,
        attempts: item.attempts + 1,
        lastAttemptAt: item.last_attempt_at ? new Date(item.last_attempt_at) : null
      };
    })();
  }

  complete(id: number): boolean {
    const stmt = this.db.prepare(`
      UPDATE queue_items
      SET status = 'completed'
      WHERE id = ? AND queue_name = ?
    `);
    const result = stmt.run(id, this.queueName);
    return result.changes > 0;
  }

  fail(id: number): boolean {
    const stmt = this.db.prepare(`
      UPDATE queue_items
      SET status = 'failed'
      WHERE id = ? AND queue_name = ?
    `);
    const result = stmt.run(id, this.queueName);
    return result.changes > 0;
  }

  retryWithDelay(id: number, delay: number): boolean {
    const scheduledAt = new Date(Date.now() + delay).toISOString();
    const stmt = this.db.prepare(`
      UPDATE queue_items
      SET status = 'pending', scheduled_at = ?
      WHERE id = ? AND queue_name = ?
    `);
    const result = stmt.run(scheduledAt, id, this.queueName);
    return result.changes > 0;
  }

  size(): number {
    const now = new Date().toISOString();
    const stmt = this.db.prepare(`
      SELECT COUNT(*) as count FROM queue_items
      WHERE queue_name = ? AND status = 'pending' AND scheduled_at <= ?
    `);
    const result = stmt.get(this.queueName, now) as { count: number };
    return result.count;
  }
} 